prov/util/src/util_av.c

```diff
@@ -269,115 +256,36 @@ static int fi_verify_av_insert(struct util_av *av, uint64_t flags)
 /*
  * Must hold AV lock
  */
-static int util_av_hash_insert(struct util_av_hash *hash, int slot,
-			       int index, int *table_slot)
-{
-	int entry, i;
-
-	if (slot < 0 || slot >= hash->slots)
-		return -FI_EINVAL;
-
-	if (hash->table[slot].index == UTIL_NO_ENTRY) {
-		hash->table[slot].index = index;
-		if (table_slot)
-			*table_slot = slot;
-		return 0;
-	}
-
-	if (hash->free_list == UTIL_NO_ENTRY)
-		return -FI_ENOSPC;
-
-	entry = hash->free_list;
-	hash->free_list = hash->table[hash->free_list].next;
-
-	for (i = slot; hash->table[i].next != UTIL_NO_ENTRY; )
-		i = hash->table[i].next;
-
-	hash->table[i].next = entry;
-	if (table_slot)
-		*table_slot = i;
-	hash->table[entry].index = index;
-	hash->table[entry].next = UTIL_NO_ENTRY;
-	return 0;
-}
-
-/* Caller must hold `av::lock` */
-static inline
-int util_av_lookup_index(struct util_av *av, const void *addr,
-			 int slot, int *table_slot)
-{
-	int i, ret = -FI_ENODATA;
-
-	if (av->hash.table[slot].index == UTIL_NO_ENTRY) {
-		FI_DBG(av->prov, FI_LOG_AV, "no entry at slot (%d)\n", slot);
-		goto out;
-	}
-
-	for (i = slot; i != UTIL_NO_ENTRY; i = av->hash.table[i].next) {
-		if (!memcmp(ofi_av_get_addr(av, av->hash.table[i].index), addr,
-			    av->addrlen)) {
-			ret = av->hash.table[i].index;
-			if (table_slot)
-				*table_slot = i;
-			FI_DBG(av->prov, FI_LOG_AV, "entry at index (%d)\n", ret);
-			break;
-		}
-	}
-out:
-	FI_DBG(av->prov, FI_LOG_AV, "%d\n", ret);
-	return ret;
-}
-
-/*
- * Must hold AV lock
- */
-int ofi_av_insert_addr(struct util_av *av, const void *addr, int slot, int *index)
+int ofi_av_insert_addr(struct util_av *av, const void *addr, fi_addr_t *fi_addr)
 {
 	struct dlist_entry *av_entry;
 	struct util_ep *ep;
 	int ret;
+	struct util_av_entry *entry = NULL;
 
-	if (OFI_UNLIKELY(av->free_list == UTIL_NO_ENTRY)) {
-		FI_WARN(av->prov, FI_LOG_AV, "AV is full\n");
-		return -FI_ENOSPC;
-	}
-
-	if (av->flags & OFI_AV_HASH) {
-		int table_slot;
-
-		if (OFI_UNLIKELY(slot < 0 || slot >= av->hash.slots)) {
-			FI_WARN(av->prov, FI_LOG_AV, "invalid slot (%d)\n", slot);
-			return -FI_EINVAL;
-		}
-		ret = util_av_lookup_index(av, addr, slot, &table_slot);
-		if (ret != -FI_ENODATA) {
-			*index = ret;
-			ofi_atomic_inc32(&av->hash.table[table_slot].use_cnt);
-			return 0;
-		}
-		ret = util_av_hash_insert(&av->hash, slot, av->free_list,
-					  &table_slot);
-		if (ret) {
-			FI_WARN(av->prov, FI_LOG_AV,
-				"failed to insert addr into hash table\n");
-			return ret;
-		}
-		ofi_atomic_inc32(&av->hash.table[table_slot].use_cnt);
+	HASH_FIND(hh, av->hash, &addr, av->addrlen, entry);
+	if (entry) {
+		*fi_addr = util_get_buf_index(av->av_entry_pool, entry);
```

Comment:
Is fi_addr guaranteed to be non-null here?

Reply:
Good catch! This is not guaranteed. Fixed this.